<a href="https://colab.research.google.com/github/Awangnugrawan/Python-Fundamental-from-MySkill/blob/main/Function.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Function

In [3]:
# function definition
def greet(name):
    ''' 
    fungsi ini menyapa
    nama orang yang masuk
    sebagai parameter 
    ''' 
    print("Hello, " + name + ". Good morning! Welcome to myskill!")

In [4]:
print(greet.__doc__) #memanggil doc string = string yang menggunakan (''')

 
    fungsi ini menyapa
    nama orang yang masuk
    sebagai parameter 
    


In [5]:
greet('zaza') #memanggil fungsi

Hello, zaza. Good morning! Welcome to myskill!


# Function Arguments

In [6]:
def greet(name, msg):  #dua parameter
    """Fungsi ini menyapa
    orang dengan pesan yang diberikan"""
    print("Hello", name + ', ' + msg)

greet("Zaza", "Good morning!")

Hello Zaza, Good morning!


## Python Default Arguments

Kita dapat memberikan nilai default ke argumen dengan menggunakan operator penugasan (=). Berikut adalah contoh.

In [8]:
def greet(name, msg="Good morning!"):
    """
    Fungsi ini menyapa
    orang dengan
    pesan yang disediakan.

    Jika pesan tidak diberikan,
    defaultnya adalah "Good 
    morning!"
    """

    print("Hello", name + ', ' + msg)


greet("Faras")
greet("Zaza", "Apa kabar")

Hello Faras, Good morning!
Hello Zaza, Apa kabar


Sejumlah argumen dalam suatu fungsi dapat memiliki nilai default. Tetapi begitu kita memiliki argumen default, semua argumen di sebelah kanannya juga harus memiliki nilai default.

## Python Keyword Arguments


Saat kita memanggil fungsi dengan beberapa nilai, nilai-nilai ini ditugaskan ke argumen sesuai dengan posisinya.

Sebagai contoh, pada fungsi di atas greet(), ketika kita menyebutnya sebagai greet("Bruce", "Bagaimana kabarmu?"), nilai "Bruce" diberikan ke name argumen dan juga "Bagaimana kabarmu? " ke msg

Python memungkinkan fungsi dipanggil menggunakan argumen kata kunci. Ketika kita memanggil fungsi dengan cara ini, urutan (posisi) argumen dapat diubah. Panggilan berikut ke fungsi di atas semuanya valid dan menghasilkan hasil yang sama.

In [7]:
# 2 keyword arguments
greet(name = "Zaza",msg = "How do you do?")

# 2 keyword arguments (out of order)
greet(msg = "How do you do?",name = "Zaza") 

# 1 positional, 1 keyword argument
greet("Zaza", msg = "How do you do?")     

Hello Zaza, How do you do?
Hello Zaza, How do you do?
Hello Zaza, How do you do?


 ## Python Arbitrary Arguments

Terkadang, kita tidak mengetahui terlebih dahulu jumlah argumen yang akan dimasukkan ke dalam suatu fungsi. Python memungkinkan kita untuk menangani situasi semacam ini melalui pemanggilan fungsi dengan sejumlah argumen yang berubah-ubah.

Dalam definisi fungsi, kita menggunakan tanda bintang (*) sebelum nama parameter untuk menunjukkan jenis argumen ini. Berikut adalah contoh.

In [9]:
def greet(*names):
    """Fungsi ini menyapa semua
    orang dalam nama tuple."""

    # names is a tuple with arguments
    for name in names:
        print("Hello", name)


greet("A", "B", "C", "D",5,7)

Hello A
Hello B
Hello C
Hello D
Hello 5
Hello 7


# Python Recursion

Berikut ini adalah contoh fungsi rekursif untuk mencari faktorial suatu bilangan bulat.

Faktorial suatu bilangan adalah hasil kali semua bilangan bulat dari 1 sampai dengan bilangan tersebut. Misalnya, faktorial dari 6 (dilambangkan dengan 6!) adalah 1x2x3x4x5x6 = 720

In [12]:
def factorial(x):
    """Ini adalah fungsi rekursif
    untuk menemukan faktorial dari bilangan bulat"""

    if x == 1:
        return 1
    else:
        return x * factorial(x-1)


num = 5
print("Nilai factorial dari", num, "adalah", factorial(num))

Nilai factorial dari 5 adalah 120


factorial(5)          # 1st call with 5\
5 * factorial(4)      # 2nd call with 4\
5 * 4 * factorial(3)  # 3rd call with 3\
5 * 4 * 3 * factorial(2) #4th call with 2\
5 * 4 * 3 * 2 * factorial(1) #5th call with 1

5 * 4 * 3 * 2 * 1     # return = 120


# Anonymous Function

In [13]:
# Program to show the use of lambda functions
double = lambda x: x * 2

print(double(5))

10


Kita menggunakan fungsi lambda ketika kita membutuhkan fungsi tanpa nama untuk waktu yang singkat.

Dalam Python, kita biasanya menggunakannya sebagai argumen ke fungsi tingkat tinggi (fungsi yang menggunakan fungsi lain sebagai argumen). Fungsi Lambda digunakan bersama dengan fungsi bawaan seperti filter(), map() dll.

Fungsi filter() dalam Python mengambil fungsi dan list sebagai argumen.

Fungsi dipanggil dengan semua item dalam list dan list baru dikembalikan yang berisi item yang fungsi dievaluasi ke True.

Berikut adalah contoh penggunaan fungsi filter() untuk memfilter hanya angka genap dari list.

In [18]:
# Program to filter out only the even items from a list
my_list = [1, 5, 4, 6, 8, 11, 3, 12]

new_list = list(filter(lambda x: (x%2 == 0) , my_list))

print(new_list)

[4, 6, 8, 12]


Fungsi map() dalam Python mengambil fungsi dan list.

Fungsi dipanggil dengan semua item dalam list dan list baru dikembalikan yang berisi item yang dikembalikan oleh fungsi itu untuk setiap item.

Berikut adalah contoh penggunaan fungsi map() untuk menggandakan semua item dalam list.

In [19]:
# Program to double each item in a list using map()

my_list = [1, 5, 4, 6, 8, 11, 3, 12]

new_list = list(map(lambda x: x * 2 , my_list))

print(new_list)

[2, 10, 8, 12, 16, 22, 6, 24]


# Global, Local and Non-Local Variable

## Global Variable and Local Variable

In [ ]:
x = "global"

def foo():
    print("x dalam:", x)


foo()
print("x luar:", x)


x dalam: global
x luar: global


Dalam kode di atas, kita membuat x sebagai variabel global dan mendefinisikan foo() untuk mencetak variabel global x. Terakhir, kita panggil foo() yang akan mencetak nilai x.

Bagaimana jika kita ingin mengubah nilai x di dalam suatu fungsi?

In [ ]:
x = "global"

def foo():
    x = x * 2
    print(x)

foo()

UnboundLocalError: ignored

Jadi pada code diatas akan menghasilkan error (local variable 'x' referenced before assignment) ini karena variabel (x) nya di definisikan ulang sebagai lokal dalam suatu function, padahal variabel (x) telah di definisikan sebagai variabel global di luar function sehingga menyebabkan error. Dan untuk cara mengatasinya menggunakan keyword **global** di dalam suatu function seperti cara di bawah ini:

In [20]:
x = "global"

def foo():
    global x
    x = x * 2
    print(x)

foo()

globalglobal


In [21]:
x = "global "

def foo():
    global x
    y = "local"
    x = x * 2
    print(x)
    print(y)

foo()

global global 
local


## Non-Local Variable

In [ ]:
def outer():
    x = "local"

    def inner():
        nonlocal x
        x = "nonlocal"
        print("inner:", x)

    inner()
    print("outer:", x)


outer()

inner: nonlocal
outer: nonlocal


# Thank you